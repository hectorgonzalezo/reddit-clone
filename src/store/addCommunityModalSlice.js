import { createSlice } from '@reduxjs/toolkit';

export const addCommunityModalSlice = createSlice({
  name: 'addCommunityModalVisible',
  initialState: false,
  reducers: {
    toggleAddCommunityModal: (state, action) => {
      return !state
    }
  },
});

export const { toggleAddCommunityModal } = addCommunityModalSlice.actions;

export const selectAddCommunityModalVisibility = (state) => state.addCommunityModalVisible;

export default addCommunityModalSlice.reducer;
