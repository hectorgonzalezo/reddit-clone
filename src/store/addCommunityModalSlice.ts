import { createSlice } from '@reduxjs/toolkit';

export const addCommunityModalSlice = createSlice({
  name: 'addCommunityModalVisible',
  initialState: false,
  reducers: {
    toggleAddCommunityModal: (state) => !state
  },
});

export const { toggleAddCommunityModal } = addCommunityModalSlice.actions;

export const selectAddCommunityModalVisibility = (state: AddCommunityModalState): boolean => state.addCommunityModalVisible;

export default addCommunityModalSlice.reducer;
