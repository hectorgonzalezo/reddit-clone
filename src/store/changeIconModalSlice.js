import { createSlice } from '@reduxjs/toolkit';

export const changeIconModalSlice = createSlice({
  name: 'changeIconModalVisible',
  initialState: false,
  reducers: {
    toggleChangeIconModal: (state, action) => {
      return !state
    }
  },
});

export const { toggleChangeIconModal } = changeIconModalSlice.actions;

export const selectChangeIconModalVisibility = (state) => state.changeIconModalVisible;

export default changeIconModalSlice.reducer;
