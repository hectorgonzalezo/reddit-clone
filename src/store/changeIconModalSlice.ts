import { createSlice } from '@reduxjs/toolkit';
import { ChangeIconModalState } from '../type';

export const changeIconModalSlice = createSlice({
  name: 'changeIconModalVisible',
  initialState: false,
  reducers: {
    toggleChangeIconModal: (state, action) => !state
  },
});

export const { toggleChangeIconModal } = changeIconModalSlice.actions;

export const selectChangeIconModalVisibility = (state: ChangeIconModalState): boolean => state.changeIconModalVisible;

export default changeIconModalSlice.reducer;
