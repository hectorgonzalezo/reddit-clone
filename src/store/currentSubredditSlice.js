import { createSlice } from '@reduxjs/toolkit';

export const currentSubredditSlice = createSlice({
  name: 'currentSubreddit',
  initialState: null,
  reducers: {
    changeCurrentSubreddit: (state, action) => {
      return action.payload
    }
  },
});

export const { changeCurrentSubreddit } = currentSubredditSlice.actions;

export const selectCurrentSubreddit = (state) => state.currentSubreddit;

export default currentSubredditSlice.reducer;
