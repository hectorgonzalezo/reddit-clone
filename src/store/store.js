import { configureStore } from '@reduxjs/toolkit';
import userReducer from './userSlice';
import subredditsReducer from './subredditsSlice';
import loginModalReducer from './loginModalSlice';
import currentSubredditReducer from './currentSubredditSlice';

const store = configureStore({
  reducer: {
    user: userReducer,
    subreddits: subredditsReducer,
    loginModalVisible: loginModalReducer,
    currentSubreddit: currentSubredditReducer,
  }
});

export default store;
