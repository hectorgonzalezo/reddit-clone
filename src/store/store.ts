import { configureStore } from '@reduxjs/toolkit';
import userReducer from './userSlice';
import postsReducer from './postsSlice';
import subredditsReducer from './subredditsSlice';
import loginModalReducer from './loginModalSlice';
import changeIconModalReducer from './changeIconModalSlice';
import addCommunityModalReducer from './addCommunityModalSlice';
import currentSubredditReducer from './currentSubredditSlice';

const store = configureStore({
  reducer: {
    user: userReducer,
    posts: postsReducer,
    subreddits: subredditsReducer,
    loginModalVisible: loginModalReducer,
    changeIconModalVisible: changeIconModalReducer,
    addCommunityModalVisible: addCommunityModalReducer,
    currentSubreddit: currentSubredditReducer,
  }
});

export default store;
