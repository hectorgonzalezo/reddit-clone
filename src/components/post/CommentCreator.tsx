import React, { useState } from 'react';
import { number, string, arrayOf, objectOf, oneOfType, array, bool, func } from 'prop-types';
import { useSelector } from 'react-redux';
import { selectUser } from '../../store/userSlice';
import { database } from '../../firebase/firebase';
import loadinIcon from '../../assets/loading.gif';
import Button from '../Button';

function CommentCreator({ commentsList, index, subreddit, postId, value, edit, reloadPost }) {
  const user = useSelector(selectUser);
  const [textContent, setTextContent] = useState(value);
  const [isLoading, setIsLoading] = useState(false);

  async function addComment() {
    setIsLoading(true);
    const copiedComments = [...commentsList];
    let indexesList;
    if (edit) {
      indexesList = index.slice(0, index.length);
      // indexesList = index;
    } else {
      indexesList = index;
    }
    // access specific location where comment should be in commentsList
    const locationToAdd = indexesList.reduce(
      (prev, curr, i) => {
        // dont acces responses if its the last elemen and it's editing the post
        if (i === indexesList.length - 1 && edit) {
          return prev[curr];
        }
        return prev[curr].responses
      },
      copiedComments
    );

    // if youre editing a previous post, just update its text content
    if (edit) {
      locationToAdd.content = textContent;
    } else {
      // add comment to that location;
      locationToAdd.push({
        content: textContent,
        timePosted: new Date().toString(),
        user: user.username,
        responses: []
      });
    }

    await database.addComment(subreddit, postId, copiedComments);
    setIsLoading(false);
    reloadPost();
  }

  return (
    <div id="create-post-area" data-testid="create-post-area">
      <p>
        Comment as <a>{user.username}</a>
      </p>
      <form action="">
        <textarea
          id="text-area"
          rows="5"
          placeholder="What are your thoughts?"
          value={textContent}
          onChange={(e) => setTextContent(e.target.value)}
        />
        <Button disabled={textContent === ''} onClick={addComment}>
          {isLoading ? (
            <img src={loadinIcon} alt="" className="loading-icon" />
          ) : (
            'Reply'
          )}
        </Button>
      </form>
    </div>
  );
}

CommentCreator.defaultProps = {
  commentsList: [],
  index: [],
  value: '',
  edit: false,
};

CommentCreator.propTypes = {
  commentsList: arrayOf(objectOf(oneOfType([string, number, array]))),
  index: arrayOf(number),
  subreddit: string.isRequired,
  postId: string.isRequired,
  value: string,
  edit: bool,
  reloadPost: func.isRequired,
};

export default CommentCreator;
