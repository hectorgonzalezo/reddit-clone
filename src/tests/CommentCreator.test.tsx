import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render, screen, container, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import store from '../store/store';
import CommentCreator from '../components/post/CommentCreator';

jest.mock('../api/users');

const icon = 'https://firebasestorage.googleapis.com/v0/b/reddit-clone-83ce9.appspot.com/o/user_icon.svg?alt=media&token=50e7a9f1-8508-4d51-aac8-4d1ed9dad7a1';


beforeAll(async() => {
  await act(async () => {
    store.dispatch({
      type: "user/addUser",
      payload: {
        user: {
        username: "juan",
        email: "mock@mock.com",
        icon,
        _id: "123456789a123456789b1234",
        },
        token: '1234701923491273401243'
      },
    });
  });
})


describe('CommentCreator', () => {
  test('Display correct user name', async () => {

    await act(async () =>
      render(
        <Provider store={store}>
          <CommentCreator subreddit="mockSubreddit" postId="123" reloadPost={() => {}} />
        </Provider>,
        { wrapper: MemoryRouter }
      )
    );

    expect(screen.getByText(/juan/)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('Disable button if textarea has no text', async () => {

    await act(async () =>
      render(
        <Provider store={store}>
          <CommentCreator subreddit="mockSubreddit" postId="123" reloadPost={() => {}} />
        </Provider>,
        { wrapper: MemoryRouter }
      )
    );

    const commentButton = screen.getByRole('button');
    const textarea = screen.getByRole('textbox');

    // Button should be disabled by default
    expect(commentButton).toHaveAttribute('disabled');

    // After typing, button should be enabled
    userEvent.type(textarea, 'text');
    expect(commentButton).not.toHaveAttribute('disabled');

    userEvent.type(textarea, '{selectall}{backspace}');
    expect(commentButton).toHaveAttribute('disabled');
  });
});
