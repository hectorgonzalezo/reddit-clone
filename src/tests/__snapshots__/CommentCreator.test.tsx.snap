// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentCreator Display correct user name 1`] = `undefined`;
